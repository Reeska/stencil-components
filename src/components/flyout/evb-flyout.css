:host {
  background-color: rgba(255, 255, 255, .8);
  display: block;
  height: 100vh;
  min-width: 15vw;
  left: 0;
  opacity: 0;
  overflow-y: auto;
  position: fixed;
  top: 0;
  transform: translateX(-100%);
  transition: transform .2s ease-in, opacity .1s ease-in;
  z-index: 100;
}

:host(.flyout--open) {
  border-right: 1px solid var(--evb-primary-color);
  opacity: 1;
  transform: translateX(0);
}

header {
  display: flex;
  justify-content: space-between;
}
.close {
  background: transparent;
  border: none;
  cursor: pointer;
  font-size: 2em;
  opacity: .8;
}
