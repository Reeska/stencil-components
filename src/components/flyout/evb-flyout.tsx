import { Component, ComponentInterface, Event, EventEmitter, h, Listen, Method, Prop } from '@stencil/core';

@Component({
  tag: 'evb-flyout',
  shadow: true,
  styleUrl: 'evb-flyout.css'
})
export class EvbFlyout implements ComponentInterface {

  @Prop({ mutable: true }) open: boolean;
  @Prop() header: string;

  @Event({ eventName: 'close' }) onClose: EventEmitter<void>;
  @Event({ eventName: 'open' }) onOpen: EventEmitter<void>;

  @Listen('keydown', { target: 'window' })
  close(event: KeyboardEvent) {
    if (this.open && event.key === 'Escape') {
      this.toggle(false);
    }
  }

  @Method()
  async toggle(forceOpen?: boolean) {
    this.open = forceOpen !== undefined ? forceOpen : !this.open;

    this.open ? this.onOpen.emit() : this.onClose.emit();
  }

  hostData() {
    return {
      class: {
        'flyout--open': this.open
      }
    };
  }

  render() {
    if (!this.open) {
      return '';
    }
    const heading = <evb-header heading='3'>{this.header}</evb-header>;

    return (
      <aside class='flyout'>
        <header>
          {heading}
          <button class='close' onClick={() => this.open = false}>&times;</button>
        </header>
        <slot />
      </aside>
    );
  }
}
