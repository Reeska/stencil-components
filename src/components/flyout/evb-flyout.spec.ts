import { EvbFlyout } from './evb-flyout';

describe('EvbFlyout', () => {
  let component: EvbFlyout;
  const closeSpy = jest.fn();
  const openSpy = jest.fn();

  beforeEach(() => {
    component = new EvbFlyout();
    component.onClose = { emit: closeSpy };
    component.onOpen = { emit: openSpy };
  });

  afterEach(() => jest.resetAllMocks());

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(component.render()).toMatchSnapshot();
  });

  it('should render host class when openend', () => {
    component.open = true;
    const classes = component.hostData();
    expect(classes.class['flyout--open']).toBe(true);
    expect(component.render()).toMatchSnapshot();
  });

  it('should not render host class when openend', () => {
    component.open = false;
    const classes = component.hostData();
    expect(classes.class['flyout--open']).toBe(false);
    expect(component.render()).toMatchSnapshot();
  });

  it('should open and emit', () => {
    expect(component.open).toBeFalsy();

    component.toggle();
    expect(component.open).toBeTruthy();
    expect(openSpy).toHaveBeenCalledTimes(1);
  });

  it('should force open and emit', () => {
    expect(component.open).toBeFalsy();

    component.toggle(true); // open
    expect(component.open).toBeTruthy();
    expect(openSpy).toHaveBeenCalledTimes(1);
  });

  it('should close and emit', () => {
    expect(component.open).toBeFalsy();

    component.toggle(); // open
    component.toggle(); // close
    expect(component.open).toBeFalsy();
    expect(closeSpy).toHaveBeenCalledTimes(1);
  });

  it('should force close and emit', () => {
    expect(component.open).toBeFalsy();

    component.toggle(false); // close
    expect(component.open).toBeFalsy();
    expect(closeSpy).toHaveBeenCalledTimes(1);
  });

  it('should close with escape key and emit', () => {
    expect(component.open).toBeFalsy();

    component.toggle(true); // open
    const event = { key: 'Escape' } as KeyboardEvent;
    component.close(event);
    expect(component.open).toBeFalsy();
    expect(closeSpy).toHaveBeenCalledTimes(1);
  });

  it('should notclose with escape key when not opened', () => {
    expect(component.open).toBeFalsy();

    const event = { key: 'Escape' } as KeyboardEvent;
    component.close(event);
    expect(component.open).toBeFalsy();
    expect(closeSpy).not.toHaveBeenCalled();
  });
});
