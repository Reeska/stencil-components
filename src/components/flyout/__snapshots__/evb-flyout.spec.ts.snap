// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvbFlyout should create 1`] = `""`;

exports[`EvbFlyout should not render host class when openend 1`] = `""`;

exports[`EvbFlyout should render host class when openend 1`] = `
Object {
  "elm": undefined,
  "ishost": false,
  "vattrs": Object {
    "class": "flyout",
  },
  "vchildren": Array [
    Object {
      "elm": undefined,
      "ishost": false,
      "vattrs": null,
      "vchildren": Array [
        Object {
          "elm": undefined,
          "ishost": false,
          "vattrs": Object {
            "heading": "3",
          },
          "vchildren": Array [
            Object {
              "vtext": "",
            },
          ],
          "vkey": undefined,
          "vname": undefined,
          "vtag": "evb-header",
          "vtext": undefined,
        },
        Object {
          "elm": undefined,
          "ishost": false,
          "vattrs": Object {
            "class": "close",
            "onClick": [Function],
          },
          "vchildren": Array [
            Object {
              "vtext": "Ã—",
            },
          ],
          "vkey": undefined,
          "vname": undefined,
          "vtag": "button",
          "vtext": undefined,
        },
      ],
      "vkey": undefined,
      "vname": undefined,
      "vtag": "header",
      "vtext": undefined,
    },
    Object {
      "elm": undefined,
      "ishost": false,
      "vattrs": null,
      "vchildren": null,
      "vkey": undefined,
      "vname": undefined,
      "vtag": "slot",
      "vtext": undefined,
    },
  ],
  "vkey": undefined,
  "vname": undefined,
  "vtag": "aside",
  "vtext": undefined,
}
`;
